call of overloaded ‘for_each_value_in_order<(nda::shape<>::rank() - 1)>(nda::shape<>::index_type, void (&)(const int&, int&), std::pair<const int*, std::tuple<> >&, std::pair<int*, std::tuple<> >&)’ is ambiguous

`mkdir -p obj/test
g++ -I. -c -o obj/test/algorithm.o test/algorithm.cpp  -O2 -ffast-math -fstrict-aliasing -march=native  -std=c++14 -Wall
In file included from test/algorithm.cpp:15:
./array.h: In instantiation of ‘void nda::for_each_value_in_order(const Shape&, const ShapeA&, PtrA, const ShapeB&, PtrB, Fn&&) [with Shape = nda::shape<>; ShapeA = nda::shape<>; PtrA = const int*; ShapeB = nda::shape<>; PtrB = int*; Fn = void (&)(const int&, int&); <template-parameter-1-7> = void]’:
./array.h:1796:28:   required from ‘static void nda::copy_shape_traits<ShapeSrc, ShapeDst>::for_each_value(const ShapeSrc&, TSrc, const ShapeDst&, TDst, Fn&&) [with Fn = void (&)(const int&, int&); TSrc = const int*; TDst = int*; ShapeSrc = nda::shape<>; ShapeDst = nda::shape<>]’
./array.h:2598:56:   required from ‘void nda::copy(const nda::array_ref<TSrc, ShapeSrc>&, const nda::array_ref<TDst, ShapeDst>&) [with TSrc = const int; TDst = int; ShapeSrc = nda::shape<>; ShapeDst = nda::shape<>; <template-parameter-1-5> = void]’
./array.h:2614:7:   required from ‘void nda::copy(const nda::array<TSrc, ShapeSrc, AllocSrc>&, nda::array<TDst, ShapeDst, AllocDst>&) [with TSrc = int; TDst = int; ShapeSrc = nda::shape<>; ShapeDst = nda::shape<>; AllocSrc = std::allocator<int>; AllocDst = std::allocator<int>; <template-parameter-1-7> = void]’
test/algorithm.cpp:76:12:   required from here
./array.h:1602:55: error: call of overloaded ‘for_each_value_in_order<(nda::shape<>::rank() - 1)>(nda::shape<>::index_type, void (&)(const int&, int&), std::pair<const int*, std::tuple<> >&, std::pair<int*, std::tuple<> >&)’ is ambiguous
 1602 |   internal::for_each_value_in_order<Shape::rank() - 1>(shape.extent(), fn, a, b);
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
./array.h:1361:41: note: candidate: ‘void nda::internal::for_each_value_in_order(const ExtentType&, Fn&&, Ptrs ...) [with long unsigned int D = 18446744073709551615; ExtentType = std::tuple<>; Fn = void (&)(const int&, int&); Ptrs = {std::pair<const int*, std::tuple<> >, std::pair<int*, std::tuple<> >}]’
 1361 | NDARRAY_INLINE NDARRAY_HOST_DEVICE void for_each_value_in_order(
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~
./array.h:1369:41: note: candidate: ‘void nda::internal::for_each_value_in_order(const std::tuple<>&, Fn&&, Ptrs ...) [with long unsigned int D = 18446744073709551615; Fn = void (&)(const int&, int&); Ptrs = {std::pair<const int*, std::tuple<> >, std::pair<int*, std::tuple<> >}]’
 1369 | NDARRAY_INLINE NDARRAY_HOST_DEVICE void for_each_value_in_order(
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~
./array.h: In instantiation of ‘void nda::for_each_value_in_order(const Shape&, Ptr, Fn&&) [with Shape = nda::shape<>; Ptr = int*; Fn = nda::generate(const nda::array_ref<T, Shape>&, Generator&&) [with T = int; Shape = nda::shape<>; Generator = int (&)(); <template-parameter-1-4> = int]::<lambda(int&)>&; <template-parameter-1-4> = void]’:
./array.h:1771:28:   required from ‘static void nda::shape_traits<Shape>::for_each_value(const Shape&, Ptr, Fn&&) [with Ptr = int*; Fn = nda::generate(const nda::array_ref<T, Shape>&, Generator&&) [with T = int; Shape = nda::shape<>; Generator = int (&)(); <template-parameter-1-4> = int]::<lambda(int&)>&; Shape = nda::shape<>]’
./array.h:1978:38:   required from ‘void nda::array_ref<T, Shape>::for_each_value(Fn&&) const [with Fn = nda::generate(const nda::array_ref<T, Shape>&, Generator&&) [with T = int; Shape = nda::shape<>; Generator = int (&)(); <template-parameter-1-4> = int]::<lambda(int&)>; <template-parameter-2-2> = void; T = int; Shape = nda::shape<>]’
./array.h:2747:3:   required from ‘void nda::generate(const nda::array_ref<T, Shape>&, Generator&&) [with T = int; Shape = nda::shape<>; Generator = int (&)(); <template-parameter-1-4> = int]’
./array.h:2752:11:   required from ‘void nda::generate(nda::array<T, Shape, Alloc>&, Generator&&) [with T = int; Shape = nda::shape<>; Alloc = std::allocator<int>; Generator = int (&)(); <template-parameter-1-5> = int]’
test/algorithm.cpp:73:19:   required from here
./array.h:1585:55: error: call of overloaded ‘for_each_value_in_order<(nda::shape<>::rank() - 1)>(nda::shape<>::index_type, nda::generate(const nda::array_ref<T, Shape>&, Generator&&) [with T = int; Shape = nda::shape<>; Generator = int (&)(); <template-parameter-1-4> = int]::<lambda(int&)>&, std::pair<int*, std::tuple<> >&)’ is ambiguous
 1585 |   internal::for_each_value_in_order<Shape::rank() - 1>(shape.extent(), fn, base_and_stride);
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./array.h:1361:41: note: candidate: ‘void nda::internal::for_each_value_in_order(const ExtentType&, Fn&&, Ptrs ...) [with long unsigned int D = 18446744073709551615; ExtentType = std::tuple<>; Fn = nda::generate(const nda::array_ref<T, Shape>&, Generator&&) [with T = int; Shape = nda::shape<>; Generator = int (&)(); <template-parameter-1-4> = int]::<lambda(int&)>&; Ptrs = {std::pair<int*, std::tuple<> >}]’
 1361 | NDARRAY_INLINE NDARRAY_HOST_DEVICE void for_each_value_in_order(
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~
./array.h:1369:41: note: candidate: ‘void nda::internal::for_each_value_in_order(const std::tuple<>&, Fn&&, Ptrs ...) [with long unsigned int D = 18446744073709551615; Fn = nda::generate(const nda::array_ref<T, Shape>&, Generator&&) [with T = int; Shape = nda::shape<>; Generator = int (&)(); <template-parameter-1-4> = int]::<lambda(int&)>&; Ptrs = {std::pair<int*, std::tuple<> >}]’
 1369 | NDARRAY_INLINE NDARRAY_HOST_DEVICE void for_each_value_in_order(
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~
./array.h: In instantiation of ‘void nda::for_each_value_in_order(const Shape&, const ShapeA&, PtrA, const ShapeB&, PtrB, Fn&&) [with Shape = nda::shape<>; ShapeA = nda::shape<>; PtrA = const int*; ShapeB = nda::shape<>; PtrB = const int*; Fn = nda::array_ref<T, Shape>::operator!=(const nda::array_ref<T, Shape>&) const [with T = const int; Shape = nda::shape<>]::<lambda(nda::array_ref<const int, nda::shape<> >::const_reference, nda::array_ref<const int, nda::shape<> >::const_reference)>&; <template-parameter-1-7> = void]’:
./array.h:1796:28:   required from ‘static void nda::copy_shape_traits<ShapeSrc, ShapeDst>::for_each_value(const ShapeSrc&, TSrc, const ShapeDst&, TDst, Fn&&) [with Fn = nda::array_ref<T, Shape>::operator!=(const nda::array_ref<T, Shape>&) const [with T = const int; Shape = nda::shape<>]::<lambda(nda::array_ref<const int, nda::shape<> >::const_reference, nda::array_ref<const int, nda::shape<> >::const_reference)>; TSrc = const int*; TDst = const int*; ShapeSrc = nda::shape<>; ShapeDst = nda::shape<>]’`

--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,4 @@
-CFLAGS := $(CFLAGS) -O2 -ffast-math -fstrict-aliasing
+CFLAGS := $(CFLAGS) -O2 -ffast-math -fstrict-aliasing -fPIE
 CXXFLAGS := $(CXXFLAGS) -std=c++14 -Wall
 LDFLAGS := $(LDFLAGS)
 

--- a/array.h
+++ b/array.h
@@ -1357,15 +1357,18 @@ NDARRAY_UNIQUE NDARRAY_HOST_DEVICE void for_each_value_in_order_impl(
   }
 }
 
-template <size_t D, class ExtentType, class Fn, class... Ptrs>
+template <size_t D, class ExtentType, class Fn, class... Ptrs,
+    std::enable_if_t<(D < std::tuple_size<ExtentType>::value), int> = 0>
 NDARRAY_INLINE NDARRAY_HOST_DEVICE void for_each_value_in_order(
     const ExtentType& extent, Fn&& fn, Ptrs... ptrs) {
   using is_inner_loop = std::conditional_t<D == 0, std::true_type, std::false_type>;
   for_each_value_in_order_impl<D>(is_inner_loop(), extent, fn, ptrs...);
 }
 
-// Scalar buffers are a special case.
-template <size_t D, class Fn, class... Ptrs>
+// Scalar buffers are a special case. The enable_if here (and above) are a workaround for a bug in
+// old versions of GCC that causes this overload to be ambiguous.
+template <size_t D, class Fn, class... Ptrs,
+    std::enable_if_t<(D == -1), int> = 0>
 NDARRAY_INLINE NDARRAY_HOST_DEVICE void for_each_value_in_order(
     const std::tuple<>& extent, Fn&& fn, Ptrs... ptrs) {
   fn(*std::get<0>(ptrs)...);
@@ -1387,20 +1390,15 @@ NDARRAY_HOST_DEVICE std::tuple<> make_compact_dims() {
 template <index_t CurrentStride, index_t Min, index_t Extent, index_t Stride, class... Dims>
 NDARRAY_HOST_DEVICE auto make_compact_dims(
     const dim<Min, Extent, Stride>& dim0, const Dims&... dims) {
-  // We already know the stride of this dimension.
-  return std::tuple_cat(std::make_tuple(dim<Min, Extent, Stride>(dim0.min(), dim0.extent())),
-      make_compact_dims<CurrentStride>(dims...));
-}
 
-template <index_t CurrentStride, index_t Min, index_t Extent, class... Dims>
-NDARRAY_HOST_DEVICE auto make_compact_dims(const dim<Min, Extent>& dim0, const Dims&... dims) {
   // If we know the extent of this dimension, we can also provide
   // a constant stride for the next dimension.
   constexpr index_t NextStride = static_mul(CurrentStride, Extent);
-  // Give this dimension the current stride, and don't give it a
-  // runtime stride. If CurrentStride is static, that will be the
-  // stride. If not, it will be dynamic, and resolved later.
-  return std::tuple_cat(std::make_tuple(dim<Min, Extent, CurrentStride>(dim0.min(), dim0.extent())),
+  // Give this dimension the current stride if we don't have one already,
+  // and don't give it a runtime stride. If CurrentStride is static, that
+  // will be the stride. If not, it will be dynamic, and resolved later.
+  constexpr index_t NewStride = is_static(Stride) ? Stride : CurrentStride;
+  return std::tuple_cat(std::make_tuple(dim<Min, Extent, NewStride>(dim0.min(), dim0.extent())),
       make_compact_dims<NextStride>(dims...));
 }
 

--- a/test/readme.cpp
+++ b/test/readme.cpp
@@ -156,8 +156,8 @@ TEST(readme_ein_reduce) {
   enum { i = 0, j = 1, k = 2, l = 3 };
 
   // Dot product dot1 = dot2 = x.y:
-  vector<float> x({10});
-  vector<float> y({10});
+  vector<float> x({10}, 0.0f);
+  vector<float> y({10}, 0.0f);
   float dot1 = make_ein_sum<float>(ein<i>(x) * ein<i>(y));
   float dot2 = 0.0f;
   ein_reduce(ein(dot2) += ein<i>(x) * ein<i>(y));
@@ -184,7 +184,7 @@ TEST(readme_ein_reduce) {
 
   // Maximum of each x-y plane of a 3D volume:
   dense_array<float, 3> T({8, 12, 20});
-  dense_array<float, 1> max_xy({20});
+  dense_array<float, 1> max_xy({20}, 0.0f);
   auto r = ein<k>(max_xy);
   ein_reduce(r = max(r, ein<i, j, k>(T)));
 

